File: ios/SecretaryKit/Sources/SecretaryKit/VaultAccess/ChangeMonitorFactory.swift
import Foundation
import SecretaryVaultAccess

/// Compose a ready-to-start `ChangeDetectionMonitor` over the real conformers for
/// `folder`. `debounceWindow` defaults to the production value; tests pass a tiny
/// window. Must be called on the main actor (the monitor is `@MainActor`).
@MainActor
public func makeChangeMonitor(
    folder: URL,
    debounceWindow: Duration = ChangeDetectionTuning.defaultDebounceWindow,
    onChange: @escaping () -> Void
) -> ChangeDetectionMonitor {
    ChangeDetectionMonitor(
        detector: FolderChangeDetector(debounceWindow: debounceWindow),
        watch: PresenterFolderWatch(folder: folder),
        scheduler: DispatchFlushScheduler(),
        onChange: onChange)
}

File: android/kit/src/main/kotlin/org/secretary/sync/ChangeMonitorFactory.kt
package org.secretary.sync

import android.os.Looper
import java.io.File
import kotlin.time.Duration

/**
 * Composes the real adapters with a fresh [FolderChangeDetector] into a ready-to-start
 * [ChangeDetectionMonitor] for [folder]. Mirror of the iOS makeChangeMonitor factory.
 * Must be called on the main thread (the returned monitor is main-thread-confined); this
 * is enforced with a fast-fail check so a background-thread misuse crashes in development
 * rather than producing a silently misconfigured monitor.
 */
fun makeChangeMonitor(
    folder: File,
    debounceWindow: Duration = ChangeDetectionTuning.defaultDebounceWindow,
    onChange: () -> Unit,
): ChangeDetectionMonitor {
    check(Looper.myLooper() == Looper.getMainLooper()) {
        "makeChangeMonitor must be called on the main thread"
    }
    return ChangeDetectionMonitor(
        detector = FolderChangeDetector(debounceWindow),
        watch = FileObserverFolderWatch(folder),
        scheduler = HandlerFlushScheduler(),
        onChange = onChange,
    )
}
